### ISSUE TITLE
Log Job Lifecycle

### ISSUE THREAD

**Issue by chrismeyersfsu** [ansible/awx#8749]
##### ISSUE TYPE
 - Feature Idea

##### SUMMARY
Users want to know why a job isn't running.

More generally, we want to tell user what and when things are happening for a Job in AWX. We can accomplish this if we know the log the lifecycle of a job.

### Job Life-cycle

Many sub-systems in AWX perform operations, both serially and in parallel, on a single Job. Each operation takes time and contributes to the total runtime of job outside of the actual playbook run. 

| State                           | Description                                                  | Reason(s)                                                    |
| ------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| pending_epoch                   | {job.type}-{job.id} created                                  | Job created in the database and has yet to be seen by the task manager |
| pending_ack                     | {job.type}-{job.id} acknowledged                             | Job seen for the first time by the task manager              |
| pending_blocked_business        | {job.type2}-{job.id2} blocked by {job.type2}-{job.id2}       | Job blocked on Project Update<br />Job blocked on Inventory Update<br />Project Update blocked on <br />Job blocked on same-job-template Job (allow_simultaneous=False)<br /><br />Project Update blocked on previous Project Update<br /><br />Inventory Update blocked on previous Inventory Update |
| pending_blocked_capacity        | {job.type}-{job.id} needs {capacity_needed}. Considered groups [({group.name[i]}, group.capacity_remaining[i]), ] | Job blocked because the amount of capacity required by the task exceeds any single group remaining capacity |
| waiting                         | {job.type}-{job.id} transitioned to waiting, will run on node {job.execution_node} and consumes capacity {job.capacity_consumed} from group {group.name} leaving the group with {group.remaining_capacity} capacity | Job is not blocked by business logic and enough capacity is available to run. |
| running_pre                     | {job.type}-{job.id} starting                                 | Job has been picked up by the dispatcher and is set from waiting to running |
| running_started                 | {job.type}-{job.id} starting playbook run                    | Job has finished pre-run checks and parameter gathering and is going to start running the playbook |
| running_finished                | {job.type}-{job.id} finished playbook run                    | Job has finished running the playbook                        |
| running_post                    | {job.type}-{job.id}                                          |                                                              |
| running_finalize                |                                                              |                                                              |
| running_finished                | {job.type}-{job.id} finish                                   | Job status set to final state. At this point, the job status is set and will not change (i.e. successful, failed, error, canceled) |
| finished_processing_facts_begin | {job.type}-{job.id} fact processing started                  |                                                              |
| finished_processing_facts_end   | {job.type}-{job.id} fact processing finished                 |                                                              |
| finished_processing_cache?      | Project Update processing cache?                             |                                                              |
| finished_processing_events      | {job.type}-{job.id} finished processing events               | Job events all persisted to the database, sent to the websocket sub-system, sent to the logging sub-system |
| finished_notifications_begin    | {job.type}-{job.id} notification sending started             | Job notification                                             |
| finished_notifications_end      | {job.type}-{job.id} notification sending finished            |                                                              |

### Recording Job Life-Cycle

Each state a job transitions to should be logged along with a timestamp. This can be accomplished in many ways, we could spend the effort to ensure the information wanted is captured in (1) log messages. After iterating the states above the system starts to look more like a tracing system than a job lifecycle log. We could bring in something like (2) OpenTracing but this would require our users to run yet another service to consume the tracing data. A middle-ground would be to (3) store the job lifecycle evnets in the database and associate them with a Job.

#### Better Log Messages
Assign a particular namespace to the job life-cycle logging i.e. `awx.main.job_lifecycle`. All state transitions noted above would flow through this logger namespace. It would be up to the user to analyze, aggregate, visualize this data. AWX could not make use of this data, nor can it surface the data to the user via the UI.

Currently, our log messages are human readable. Instead, for this feature, the log messages should be structured.

#### Tracing
Use OpenTracing to log job life-cycle. The "trouble" here is not the tracing but using the tracing data. Picking, shipping, and requiring our users to manage a distributed tracing server (i.e. Jaeger) is a big ask.

#### Store it in Postgres
Creating a new Model and adding a Many2Many Relationship from Jobs to job life-cycle logs allows for AWX to make use of the data. The biggest concern here is performance/management of this log.

### What can we do with this data?
Most simplye, the data can be presented in the UI as a per-job log so that users can know the answer to "why isn't my job running?" or "Why isn't my playbook running?"

A more complex view of the data might be shown as a timeline or gannt chart. This would present the data across multiple jobs. Parallelism, capacity consumption and availability, and task manager performance are example metrics that can be derived from aggregating this data.

### Misc questions
* Should we always fill a state entry in as if we progressed through the state? I'm thinking about canceled. You skip a bunch of states, but it would simplify the decision tree/state diagram.
  * No, because entering some states is conditional. For example, entering the pending_blocked_business state communicates that the job was blocked. Often, it's the case that a job never enters this state.

**Comment by fosterseth** [comment]
@chrismeyersfsu 
The task manager (and maybe other pieces) runs in a cycle and processes the same task on each cycle. So a logger-only implementation would need to have a way to prevent duplicate messages. E.g. if job 99 is blocked, it might be blocked for the same reason on the next task manager cycle.

given the above, maybe a centralized, persistent system is best (write task state to db table)

there might be way to have both logging + db table. It would be nice to have these task state information alongside the rest of our logging messages, for a more detailed debugging process.

also, what do you have in mind for `job type` - do you mean `task.unified_job_template.name`? or generic `job`, `project-update`, `system-job` etc.

**Comment by chrismeyersfsu** [comment]
Generic: `job`, `project-update`, etc.

You hit the nail on the head with needing state to be able to output some of these messages. I'd say to delay that as long as possible and see how much value we get out of the messages that we can support without adding state tracking.

Thinking more on maintaining state. @fosterseth you've shown that the individual job `.save()` are expensive when the task manager decides to run jobs. Note that this is a Django overhead and not an SQL overhead, we are in a transaction, the actual SQL execution is delayed until the end, yet you showed the `.save()` was a large factor in the function level tracing. This is all to say that if we track a per-job state we must be careful as it will have performance implications.

One path forward is to log the job life-cycle points that we can without tracking state. Then in a second pass, when the performance stats feature is in place, we add in state tracking so that we have a before and after.

Another path forward could be that you go ahead and implement the state tracking, ensuring that performance is not degraded. This work will require identifying key performance data in the task manager. This data can be added to the metrics feature and surfaced for future monitoring.

**Comment by kdelee** [comment]
We've been running integration and because @fosterseth did a great job of getting the tests green before merging, have no regressions since the merge of this feature.

Additionally, in any SOS report from a test run we can see that we now have the new log file with the logs:
```
{"type": "job", "task_id": 3670, "state": "event_processing_finished", "time": "2021-02-10T12:27:04.867683+00:00"}
{"type": "job", "task_id": 3667, "state": "event_processing_finished", "time": "2021-02-10T12:27:05.405113+00:00"}
{"type": "job", "task_id": 3668, "state": "event_processing_finished", "time": "2021-02-10T12:27:05.406554+00:00"}
{"type": "job", "task_id": 3670, "state": "event_processing_finished", "time": "2021-02-10T12:27:05.407549+00:00"}
{"type": "job", "task_id": 3667, "state": "event_processing_finished", "time": "2021-02-10T12:27:05.626205+00:00"}
{"type": "job", "task_id": 3668, "state": "event_processing_finished", "time": "2021-02-10T12:27:05.628342+00:00"}
{"type": "job", "task_id": 3670, "state": "event_processing_finished", "time": "2021-02-10T12:27:05.629759+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "created", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:14.850225+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "acknowledged", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:15.058856+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "waiting", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:15.110525+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "pre_run", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:15.257831+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "preparing_playbook", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:15.261912+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "running_playbook", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:18.850556+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "post_run", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:18.853304+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "finalize_run", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:18.974692+00:00"}
{"type": "projectupdate", "task_id": 3671, "state": "event_processing_finished", "template_name": "Project - LeagueSafety\ufffd", "time": "2021-02-10T12:27:20.296109+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "created", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:26.557949+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "acknowledged", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:26.753306+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "waiting", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:26.809767+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "pre_run", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:27.206315+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "preparing_playbook", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:27.213103+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "running_playbook", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:32.218638+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "post_run", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:32.221012+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "finalize_run", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:32.405533+00:00"}
{"type": "projectupdate", "task_id": 3672, "state": "event_processing_finished", "template_name": "AWX-Collection-tests-tower_inventory_source_update-project-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:32.501646+00:00"}
{"type": "inventoryupdate", "task_id": 3673, "state": "created", "template_name": "AWX-Collection-tests-tower_inventory_source_update-source2-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:41.022908+00:00"}
{"type": "inventoryupdate", "task_id": 3673, "state": "acknowledged", "template_name": "AWX-Collection-tests-tower_inventory_source_update-source2-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:41.201302+00:00"}
{"type": "inventoryupdate", "task_id": 3673, "state": "waiting", "template_name": "AWX-Collection-tests-tower_inventory_source_update-source2-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:41.253650+00:00"}
{"type": "inventoryupdate", "task_id": 3673, "state": "pre_run", "template_name": "AWX-Collection-tests-tower_inventory_source_update-source2-MqYYNqREjKgKfGwr", "time": "2021-02-10T12:27:41.392351+00:00"}
```

Finally, while working on the logrotate feature, we've confirmed that this plays well with all the log rotation logic.

Thanks! Closing
